File: solutions_2453486_1/C++/swift90/tttt.cpp
#include<cstdio>
const int N=4;
char nr[5][10];
int check(char X){
	int ok;
	for(int i=0;i<N;++i){
		ok=1;
		for(int j=0;j<N;++j)
			if(nr[i][j]!=X)
				ok=0;
		if(ok)
			return 1;
	}
	for(int j=0;j<N;++j){
		ok=1;
		for(int i=0;i<N;++i)
			if(nr[i][j]!=X)
				ok=0;
		if(ok)
			return 1;
	}
	ok=1;
	for(int i=0;i<N;++i)
		if(nr[i][i]!=X)
			ok=0;
	if(ok)
		return 1;
	ok=1;
	for(int i=0;i<N;++i)
		if(nr[i][N-i-1]!=X)
			ok=0;
	return ok;
}
void solve(int test){
	int i,j;
	int xt,yt;
	xt=yt=4;
	int cnt=0;
	for(i=0;i<N;++i){
		for(j=0;j<N;++j){
			if(nr[i][j]=='T'){
				xt=i;
				yt=j;
			}
			if(nr[i][j]=='.')
				++cnt;
		}
	}
	nr[xt][yt]='X';
	if(check('X')){
		printf("Case #%d: X won\n",test);
		return;
	}
	nr[xt][yt]='O';
	if(check('O')){
		printf("Case #%d: O won\n",test);
		return;
	}
	if(!cnt){
		printf("Case #%d: Draw\n",test);
		return;
	}
	printf("Case #%d: Game has not completed\n",test);
}
int main(){
	freopen("ttt.in","r",stdin);
	freopen("ttt.out","w",stdout);
	int T;
	scanf("%d",&T);
	for(int test=1;test<=T;++test){
		fgets(nr[0],10,stdin);
		for(int i=0;i<N;++i)
			fgets(nr[i],10,stdin);
		solve(test);
	}
	fclose(stdin);
	fclose(stdout);
	return 0;
}

File: solutions_2453486_1/Python/pvncad/tttt.py
#/usr/local/bin/python

import sys

T = int(sys.stdin.readline())

players = ["X", "O"]
dirs = ["R", "C", "D"]

def solve(board, direction):
    start = [ (x, 0) for x in xrange(4) ]
    dirx  = 0
    diry  = 1

    if direction == 2:
        start = [(0, 0), (0, 3)]
        dirx = diry = 1
    elif direction == 1:
        start = [ (0, x) for x in xrange(4) ]
        dirx  = 1
        diry  = 0

    is_empty = False

    for p in start:
        char = None
        found = True 
        line = ""

        if direction == 2 and p[1] == 3:
            diry = -1

        for i in xrange(4):
            ix = (p[0] + i * dirx) % 4
            iy = (p[1] + i * diry) % 4
            line += board[ix][iy]
            if board[ix][iy] == '.':
                found = False
                is_empty = True
                break
            elif board[ix][iy] != 'T':
                if char is None:
                    char = board[ix][iy]
                elif char != board[ix][iy]:
                    found = False


        if found:
            if char == "X":
                return 1
            else:
                return 2
    if is_empty:
        return -1

    return 0


for caseno in xrange(T):
    board = [ ]

    if caseno:
        sys.stdin.readline()

    for x in xrange(4):
        board.append(list(sys.stdin.readline().strip()))

    answer = None
    for x in xrange(3):
        new_ans = solve(board, x)

        if new_ans > 0:
            answer = new_ans
            break;
        elif new_ans == -1:
            answer = new_ans

    res = "Draw"

    if answer > 0:
        res = "%s won" % (players[answer - 1])
    elif answer == -1:
        res = "Game has not completed"

    print "Case #%d: %s" % (caseno + 1, res)
